File: DSA_with_python/121.best-time-to-buy-and-sell-stock.py
#
# @lc app=leetcode id=121 lang=python3
#
# [121] Best Time to Buy and Sell Stock
#

# @lc code=start



class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        min_price = prices[0]
        max_profit = 0
        for price in prices:
            if price < min_price:
                min_price = price
            elif price - min_price > max_profit:
                max_profit = price - min_price
        return max_profit
# @lc code=end

# Example usage:
solution = Solution()
print(solution.maxProfit([7, 1, 5, 3, 6, 4]))  # Output: 5
print(solution.maxProfit([7, 6, 9, 3, 1]))    # Output: 2

File: DSA_with_c++/121.best-time-to-buy-and-sell-stock.cpp
/*
 * @lc app=leetcode id=121 lang=cpp
 *
 * [121] Best Time to Buy and Sell Stock
 */
#include <vector>
#include <climits>
using namespace std;
// @lc code=start
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int min_price = INT_MAX;
        int max_profit = 0;

        for (int price : prices) {
            if (price < min_price) {
                min_price = price; // Update the minimum price
            } else if (price - min_price > max_profit) {
                max_profit = price - min_price; // Update the maximum profit
            }
        }

        return max_profit; // Return the maximum profit found
        
    }
};
// @lc code=end

